controllers/che/checluster_controller.go

```diff
@@ -810,3 +496,339 @@ func (r *CheClusterReconciler) GetCR(request ctrl.Request) (instance *orgv1.CheC
+func reconcileIdentityProvider(deployContext *deploy.DeployContext) (*ctrl.Result, error) {
+	deleted, _ := identity_provider.ReconcileIdentityProvider(deployContext)
+	if deleted {
+		// ignore error
+		deploy.DeleteFinalizer(deployContext, deploy.OAuthFinalizerName)
+		for {
+			deployContext.CheCluster.Status.OpenShiftoAuthProvisioned = false
+			if err := deploy.UpdateCheCRStatus(deployContext, "status: provisioned with OpenShift identity provider", "false"); err != nil &&
+				errors.IsConflict(err) {
+				deploy.ReloadCheClusterCR(deployContext)
+				continue
+			}
+			break
+		}
+		for {
+			deployContext.CheCluster.Spec.Auth.OAuthSecret = ""
+			deployContext.CheCluster.Spec.Auth.OAuthClientName = ""
+
+			if err := deploy.UpdateCheCRStatus(deployContext, "clean oAuth secret name and client name", ""); err != nil &&
+				errors.IsConflict(err) {
+				deploy.ReloadCheClusterCR(deployContext)
+				continue
+			}
+			break
+		}
+	}
+
+	return nil, nil
+}
+
+func deleteOAuthInitialUserIfNeeded(deployContext *deploy.DeployContext) (*ctrl.Result, error) {
+	if util.IsOpenShift4 && util.IsDeleteOAuthInitialUser(deployContext.CheCluster) {
+		userHandler := NewOpenShiftOAuthUserHandler(deployContext.ClusterAPI.NonCachedClient)
+		if err := userHandler.DeleteOAuthInitialUser(deployContext); err != nil {
+			logrus.Errorf("Unable to delete initial OpenShift OAuth user from a cluster. Cause: %s", err.Error())
+
+			// don't try one more time
+			deployContext.CheCluster.Spec.Auth.InitialOpenShiftOAuthUser = nil
+			err := deploy.UpdateCheCRSpec(deployContext, "initialOpenShiftOAuthUser", "nil")
+			return &reconcile.Result{}, err
+		}
+
+		deployContext.CheCluster.Spec.Auth.OpenShiftoAuth = nil
+		deployContext.CheCluster.Spec.Auth.InitialOpenShiftOAuthUser = nil
+		updateFields := map[string]string{
+			"openShiftoAuth":            "nil",
+			"initialOpenShiftOAuthUser": "nil",
+		}
+
+		if err := deploy.UpdateCheCRSpecByFields(deployContext, updateFields); err != nil {
+			return &reconcile.Result{}, err
+		}
+
+		return nil, nil
+	}
+
+	// Update status if OpenShift initial user is deleted (in the previous step)
+	if deployContext.CheCluster.Spec.Auth.InitialOpenShiftOAuthUser == nil && deployContext.CheCluster.Status.OpenShiftOAuthUserCredentialsSecret != "" {
+		secret := &corev1.Secret{}
+		exists, err := getOpenShiftOAuthUserCredentialsSecret(deployContext, secret)
+		if err != nil {
+			// We should `Requeue` since we deal with cluster scope objects
+			return &ctrl.Result{RequeueAfter: time.Second}, err
+		} else if !exists {
+			deployContext.CheCluster.Status.OpenShiftOAuthUserCredentialsSecret = ""
+			if err := deploy.UpdateCheCRStatus(deployContext, "openShiftOAuthUserCredentialsSecret", ""); err != nil {
+				return &reconcile.Result{}, err
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+func doneToResult(done bool, forceRequeueIfNotDone bool, err error) (*ctrl.Result, error) {
```

Comment:
Do we really need this method?
Please consider alternatives:
- Detect reconcile status based on `done` and `error` as, for example, [here](https://github.com/eclipse-che/che-operator/blob/main/controllers/checlusterbackup/checlusterbackup_controller.go#L181)
- Return `ctrl.Result` in place